File: FirstAkkaApplication/src/main/java/akka/first/app/mapreduce/scala/actors/AggregateActor.scala
package akka.first.app.mapreduce.scala.actors

import scala.collection.immutable.Map
import scala.collection.mutable.HashMap

import akka.actor.Actor
import akka.first.app.mapreduce.scala.{ReduceData, Result}

class AggregateActor extends Actor {
  val finalReducedMap = new HashMap[String, Int]

  def receive: Receive = {
    case ReduceData(reduceDataMap) =>
      aggregateInMemoryReduce(reduceDataMap)
    case Result =>
      sender ! finalReducedMap.toString()
  }

  def aggregateInMemoryReduce(reducedList: Map[String, Int]): Unit = {
    for ((key, value) <- reducedList) {
      if (finalReducedMap contains key)
        finalReducedMap(key) = (value + finalReducedMap.get(key).get)
      else
        finalReducedMap += (key -> value)
    }
  }
}

File: FirstAkkaApplication/src/main/java/akka/first/app/mapreduce/java/actors/AggregateActor.java
package akka.first.app.mapreduce.java.actors;

import akka.actor.UntypedActor;
import akka.first.app.mapreduce.java.message.ReduceData;
import akka.first.app.mapreduce.java.message.Result;

import java.util.HashMap;
import java.util.Map;

public class AggregateActor extends UntypedActor {

    private Map<String, Integer> finalReducedMap = new HashMap<String, Integer>();

    public void onReceive(Object message) throws Exception {
        if (message instanceof ReduceData) {
            ReduceData reduceData = (ReduceData) message;
            aggregateInMemoryReduce(reduceData.getReduceDataList());
        } else if (message instanceof Result) {
            getSender().tell(finalReducedMap.toString());
        } else {
            unhandled(message);
        }
    }

    private void aggregateInMemoryReduce(Map<String, Integer> reducedList) {
        Integer count = null;
        for (String key : reducedList.keySet()) {
            if (finalReducedMap.containsKey(key)) {
                count = reducedList.get(key) + finalReducedMap.get(key);
                finalReducedMap.put(key, count);
            } else {
                finalReducedMap.put(key, reducedList.get(key));
            }
        }
    }
}
